### DISCUSSION TITLE
How to add a custom field on query result?

### DISCUSSION THREAD

**Discussion by arichyx** [prisma/prisma#24931] [Q&A]
### Question

I have an implicit many-to-many relation: User and Post. A user could like many posts, and a post could be liked by many users.

Here is the schema:
```typescript
model User {
  id         String    @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  posts      Post[]    @relation("author")
  likes      Post[]    @relation("like")

  @@map("users")
}

model Post {
  id         String    @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  author     User      @relation("author", fields: [author_id], references: [id])
  author_id  String    @db.Uuid
  comments   Comment[]
  likedBy    User[]    @relation("like")

  @@map("posts")
}
```
Now I want to get the post lists to render. For every post, there will be a heart showing if it's already liked by current user. How could I get a structure with type `Post & { isLiked: boolean }` ? It would be perfect if it could be done in one query.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @Arichy 👋 

Thank you for raising this question.

You can fetch the `post` and compute the `isLike` property using applications logic. For example:

```ts
const userId = 'currentUserId'; // Replace with the actual current user ID

const posts = await prisma.post.findMany({
  include: {
    likedBy: {
      select: {
        id: true,
      },
    },
  },
});

const postsWithIsLiked = posts.map(post => ({
  ...post,
  isLiked: post.likedBy.some(user => user.id === userId),
}));

console.log(postsWithIsLiked);
```
This approach ensures that you fetch all necessary data in one query and then compute the isLiked property in your application logic. 

You can also leverage Prisma Client Extension [computed fields](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-computed-fields) for it.

**Reply by apudiu** [reply]
This is not a solution at all. Imagine this post has 700k likes. This is also `n+1` problem. This is a NO-GO option. I recommend you @Arichy cancel marking this as an answer. If Prisma can't do the thing in the question then it's okay that prisma can't do that.

**Reply by arichyx** [reply]
@apudiu You're right. Actually I've adopted another method, but I forgot to attach it. Please refer to it below.

**Comment by arichyx** [comment]
Selected method by me: use `likedBy: { where: { id: userId }, select: { id: true } }` in `include`
If current user has liked the post, the `likedBy` array will contain 1 record, which is exactly the userId.

```typescript
 const list = await prisma.post.findMany({
    take: PAGE_SIZE,
    orderBy: { created_at: 'desc' },
    include: {
      author: {
        select: {
          id: true,
          email: true,
          name: true,
        },
      },
      likedBy: {
        where: {
          id: userId,
        },
        select: { id: true },
      },
      _count: {
        select: {
          likedBy: true,
        },
      },
    },
  });

  const populatedList = await Promise.all(
    list.map(async post => {
      const { likedBy, _count, ...rest } = post;
      return Object.assign({}, rest, {
        isLiked: userId ? likedBy.length > 0 : false,
        likeCount: _count.likedBy,
      });
    })
  );

  return json(populatedList);
```
